proof/refine/ARM/StateRelation.thy

```diff
@@ -880,33 +891,57 @@ lemma sc_replies_relation_prev_list:
   apply (clarsimp simp: sc_of_def opt_map_def projectKO_opt_sc split: option.splits)
   done
 
-(* FIXME: can generalise in terms of "sym heaps", maybe later *)
-lemma sym_refs_prev_eq_next_sc_replies:
-  "\<lbrakk> sym_refs (list_refs_of_replies' s');
-     heap_list (replyPrevs_of s') (scReply sc') replies\<rbrakk>
-    \<Longrightarrow> heap_list (replyNexts_of s') (scReply sc') (rev replies)"
-  oops
-
-lemma sc_replies_relation_next_list:
-  "\<lbrakk> sc_replies_relation s s'; sym_refs (list_refs_of_replies' s');
-     kheap s x = Some (kernel_object.SchedContext sc n);
-     ksPSpace s' x = Some (KOSchedContext sc')\<rbrakk>
-    \<Longrightarrow> heap_list (replyNexts_of s') (scReply sc') (rev (sc_replies sc))"
-  apply (frule (2) sc_replies_relation_prev_list)
-  apply (case_tac "sc_replies sc"; simp)
-  apply (clarsimp simp: sc_replies_relation_def sc_replies_of_scs_def scs_of_kh_def map_project_def)
-  apply (drule_tac x=x and y="sc_replies sc" in spec2)
-  apply (clarsimp simp: sc_of_def opt_map_def projectKO_opt_sc)
-  oops
-
-lemma ghost_relation_typ_at:
-  "ghost_relation (kheap s) ups cns \<equiv>
-   (\<forall>a sz. data_at sz a s = (ups a = Some sz)) \<and>
-   (\<forall>a n. typ_at (ACapTable n) a s = (cns a = Some n))"
-   apply (rule eq_reflection)
-   apply (clarsimp simp: ghost_relation_def typ_at_eq_kheap_obj data_at_def)
-   apply (intro conjI impI iffI allI; force)
-   done
+lemma list_refs_of_replies'_reftype[simp]:
+  "(p, reftype) \<in> list_refs_of_replies' s' p' \<Longrightarrow> reftype \<in> {ReplyPrev, ReplyNext}"
+  by (clarsimp simp: list_refs_of_replies'_def list_refs_of_reply'_def get_refs_def2
+              split: option.split_asm)
+
+(* FIXME RT: move *)
+lemma opt_map_left_Some:
+  "f x = Some y \<Longrightarrow> (f |> g) x = g y"
+  by (clarsimp simp: opt_map_def)
+
+lemma opt_map_left_None:
+  "f x = None \<Longrightarrow> (f |> g) x = None"
+  by (clarsimp simp: opt_map_def)
+
+lemma case_opt_map_distrib:
+  "((\<lambda>s. case_option None g (f s)) |> h)
+   = ((\<lambda>s. case_option None (g |> h) (f s)))"
+  by (fastforce simp: opt_map_def split: option.splits)
+
+lemma opt_map_assoc:
+  "f |> g |> h = f |> (g |> h)"
```

Comment:
Nice! We should put this the other way around, to normalise to the non-parenthesis form.